storage/rocksdb/rdb_locking_iter.h

```diff
@@ -0,0 +1,222 @@
+
+      if (my_core::thd_killed(current_thd)) {
+        m_status = rocksdb::Status::Aborted();
+        m_valid  = false; 
+        return;
+      }
+
+      const int inv = forward ? 1 : -1;
+      auto cmp= m_cfh->GetComparator();
+
+      auto end_key = m_iter->key();
+      bool endp_arg= m_is_rev_cf;
+
+      if (m_have_locked_until &&
+          cmp->Compare(end_key, rocksdb::Slice(m_locked_until))*inv <= 0) {
+        // We've already locked this range. The following has happened:
+        // - m_iter->key() returned $KEY
+        // - we got a range lock on [range_start, $KEY]
+        // - other transaction(s) have inserted row $ROW before the $KEY.
+        // - we've read $ROW and returned.
+        // Now, we're looking to lock [$ROW, $KEY] but we don't need to,
+        // we already have a lock on this range.
+      } else {
+        if (forward) {
+          m_status = m_txn->GetRangeLock(m_cfh,
+                                       rocksdb::Endpoint(target, endp_arg),
+                                       rocksdb::Endpoint(end_key, endp_arg));
+        } else {
+          m_status = m_txn->GetRangeLock(m_cfh,
+                                       rocksdb::Endpoint(end_key, endp_arg),
+                                       rocksdb::Endpoint(target, endp_arg));
+        }
+
+        // Save the bound where we locked until:
+        m_have_locked_until= true;
+        m_locked_until.assign(end_key.data(), end_key.size());
+        if (!m_status.ok()) {
+          // Failed to get a lock (most likely lock wait timeout)
+          m_valid = false;
+          return;
+        }
+        if (m_lock_count)  (*m_lock_count)++;
+      }
+
+      std::string end_key_copy= end_key.ToString();
+
+      //Ok, now we have a lock which is inhibiting modifications in the range
+      // Somebody might have done external modifications, though:
+      //  - removed the key we've found
+      //  - added a key before that key.
+
+      // First, refresh the iterator:
+      delete m_iter;
+      m_iter = m_txn->GetIterator(m_read_opts, m_cfh);
+
+      // Then, try seeking to the same row
+      if (forward)
+        m_iter->Seek(target);
+      else
+        m_iter->SeekForPrev(target);
+
+
+      if (call_next && m_iter->Valid() && !cmp->Compare(m_iter->key(), target)) {
+        if (forward)
+          m_iter->Next();
+        else
+          m_iter->Prev();
+      }
+
+      if (m_iter->Valid()) {
+        if (cmp->Compare(m_iter->key(), rocksdb::Slice(end_key_copy))*inv <= 0) {
+          // Ok, the found key is within the range.
+          m_status = rocksdb::Status::OK();
+          m_valid= true;
+          break;
+        } else {
+          // We've got a key but it is outside the range we've locked.
+          // Re-try the lock-and-read step.
+          continue;
+        }
+      } else {
+        // There's no row (within the iterator bounds perhaps). Exit now.
```

Comment:
It seems like we may need range lock between the target slice and the end range of the iterator bounds.

Reply:
Yes, done.

Reply:
Wait, the fix fix crashes with `rocksdb_enable_iterate_bounds=off` . Will fix

Reply:
Now done.